Check for server message on HTTP response error

ecederstrand opened the pull request:
If available, include this message in a custom exception class and throw that instead.

Before this PR:
```python
>>> charts.find(xxx="foo")
[...]
HTTPError: 400 Client Error: BAD REQUEST for url: http://example.com/api/v1/chart/?q=...
```

After this PR:
```python
>>> charts.find(xxx="foo")
[...]
ServerError: Filter column: xxx not allowed to filter
```

Fixes #26

opus-42 commented on supersetapiclient/base.py:
```diff
@@ -21,6 +21,18 @@ def default_string():
+            raise e
+        raise ServerError(*e.args, request=e.request, response=e.response, message=error_msg) from None
+
+
```
Let me propose another logic for this @ecederstrand. Here is what we could do:
- avoid the try... except mechanisms. The `except HTTPError` is too wide.
- do an `if ... else` using `response.status_code`. If this is 400, then we can parse answer and implement a better server.

ecederstrand replied:
I was trying to avoid diving into the logic of `response.raise_for_status()` and speculate about which `status_code` values trigger an error. I think catching `HTTPError` should be OK because we only re-raise it as something else if `response.json()["message"]` succeeds. Otherwise, it is re-raised unaltered.

ecederstrand replied:
I'm expanding the test suite and encountering other HTTP 4xx errors. So I think the above approach with less assumptions about the original exception instance is better.

opus-42 commented on supersetapiclient/exceptions.py:
```diff
@@ -1,5 +1,15 @@
     pass
+
+
+class ServerError(HTTPError):
```
Could we find another naming ? ServerError suggests there is something wrong happening on Superset Server side, which is not the case. This should be more a `BadRequestError` or something similar (i.e. request was incorrect)

ecederstrand replied:
I agree with the unclear naming. The error is not necessarily caused by a bad request, but since that's how we use it now, I think `BadRequestError` is a good name.

ecederstrand replied:
Do you want me to find a different name for `ComplexServerError` as well? `ComplexBadRequestError`? I can't come up with any non-horrible names right now...

ecederstrand replied:
Settling in `ComplexBadRequestError` in lack of better suggestion.

Commit:
Include server message on HTTP response error

If a message was returned by the server, raise a custom exception class with this message. Fixes #26

ecederstrand committed:
Merge branch 'opus-42:develop' into 26-include-server-message-in-exception

ecederstrand commented:
Since https://gitlab.com/adamatics/python/adalib/-/commit/d2f9bfcca248c59bc4864b696933cca9e470bd0f we're installing a from a branch that has the fix. I'll keep this issue open a bit until the upstream PR has been merged.

ecederstrand commented:
An alternative to this approach is to create a custom `requests.Response` class with a custom `raise_for_status()` method, and then provide a custom `requests.HTTPAdapter` implementation for the session that overrides `HTTPAdapter.build_response()` to return the custom `Response` class. The principle is described in https://stackoverflow.com/questions/42365096/sensible-way-to-extend-requests-response-class

I think the approach in this MR is less complicated and easier to follow.

ecederstrand committed:
Support server errors with a {'errors': ...} structure

ecederstrand committed:
Merge branch '26-include-server-message-in-exception' of github.com:ecederstrand/superset-api-client into 26-include-server-message-in-exception

ecederstrand commented:
Added extra exception class that turned out to be useful in https://github.com/opus-42/superset-api-client/pull/33

opus-42 commented:
Hello @ecederstrand , i started a review a few days ago. By any chance had you the chance to take a look at it ?

Best regards

ecederstrand commented:
@opus-42 It's not clear to me what you mean? I don't see any review comments to this PR.

opus-42 reviewed (changes requested):
Here is a couple changes I propose

opus-42 commented:
> @opus-42 It's not clear to me what you mean? I don't see any review comments to this PR.

Ok my bad, I think the review was not shared. Now it should be.

ecederstrand committed:
Improve naming of exception classes

ecederstrand committed:
Merge branch 'opus-42:develop' into 26-include-server-message-in-exception

ecederstrand committed:
Merge branch 'develop' into 26-include-server-message-in-exception

ecederstrand committed:
Merge branch 'opus-42:develop' into 26-include-server-message-in-exception
